src/execute.rs

```diff
@@ -0,0 +1,654 @@
+    let p_upper = tick_to_price(upper_tick)?;
+    let p_current = tick_to_price(current_tick)?;
+    let delta_x = Decimal256::one()
+        .checked_mul(p_upper.sqrt().checked_sub(p_current.sqrt())?)?
+        .checked_div(p_upper.sqrt().checked_mul(p_current.sqrt())?)?;
+
+    Ok(delta_x)
+}
+
+pub fn calc_amount_1_one_unit_liq(
+    lower_tick: i64,
+    current_tick: i64,
+) -> Result<Decimal256, ContractError> {
+    let p_lower = tick_to_price(lower_tick)?;
+    let p_current = tick_to_price(current_tick)?;
+    let delta_y = Decimal256::one().checked_mul(p_current.sqrt().checked_sub(p_lower.sqrt())?)?;
+
+    Ok(delta_y)
+}
+
+pub fn calc_asset_ratio_from_ticks(
+    upper_tick: i64,
+    current_tick: i64,
+    lower_tick: i64,
+) -> Result<(Decimal256, Decimal256), ContractError> {
+    if upper_tick < current_tick {
+        return Ok((Decimal256::zero(), Decimal256::one()));
+    }
+    if lower_tick > current_tick {
+        return Ok((Decimal256::one(), Decimal256::zero()));
+    }
+    let delta_x = calc_amount_0_one_unit_liq(upper_tick, current_tick)?;
+    let delta_y = calc_amount_1_one_unit_liq(lower_tick, current_tick)?;
+    let total_delta = delta_x.checked_add(delta_y)?;
+    let asset0_ratio = delta_x.checked_div(total_delta)?;
+    let asset1_ratio = Decimal256::one().checked_sub(asset0_ratio)?;
+
+    Ok((asset0_ratio, asset1_ratio))
+}
+
+fn get_next_sqrt_price_from_amount0_in_round_up(
+    liquidity: Decimal256,
+    sqrt_price_current: Decimal256,
+    token_in: Decimal256,
+) -> Decimal256 {
+    let numerator = liquidity * sqrt_price_current;
+    let denominator = liquidity + (token_in * sqrt_price_current);
+    numerator / denominator
+}
+
+fn get_next_sqrt_price_from_amount1_in_round_down(
+    liquidity: Decimal256,
+    sqrt_price_current: Decimal256,
+    token_in: Decimal256,
+) -> Decimal256 {
+    sqrt_price_current + (token_in / liquidity)
+}
+
+fn calc_amount_one_delta(
+    liquidity: Decimal256,
+    mut sqrt_price_current: Decimal256,
+    mut sqrt_price_next: Decimal256,
+    should_round_up: bool,
+) -> Decimal256 {
+    if sqrt_price_next > sqrt_price_current {
+        std::mem::swap(&mut sqrt_price_current, &mut sqrt_price_next);
+    }
+    let result = liquidity * (sqrt_price_current.abs_diff(sqrt_price_next));
+    match should_round_up {
+        true => result.ceil(),
+        false => result,
+    }
+}
+
+fn calc_amount_zero_delta(
+    liquidity: Decimal256,
+    mut sqrt_price_current: Decimal256,
+    mut sqrt_price_next: Decimal256,
+    should_round_up: bool,
+) -> Decimal256 {
+    if sqrt_price_next > sqrt_price_current {
+        std::mem::swap(&mut sqrt_price_current, &mut sqrt_price_next);
+    }
+    let mul1 = liquidity * (sqrt_price_current - sqrt_price_next);
+    let result = mul1 / (sqrt_price_current * sqrt_price_next);
+    match should_round_up {
+        true => result.ceil(),
+        false => result,
+    }
+}
+
+fn iterative_naive_approach(
```

Comment:
nit: instead of checking "if token_0_in" twice, suggest doing that once and extracting the common logic into a separate helper. That would help with readability